epoxy-paging/src/main/java/com/airbnb/epoxy/paging/PagedListModelCache.kt

```diff
@@ -31,130 +32,222 @@ import java.util.concurrent.Executor
+            }
+
+            // we have to reply on this private API, otherwise, paged list might be changed when models are being built,
+            // potentially creating concurrent modification problems.
+            builder.setMainThreadExecutor { runnable: Runnable ->
+                modelBuildingHandler.post(runnable)
+            }
+        }.build()
+    ) {
+        init {
+            if (modelBuildingHandler != EpoxyController.defaultModelBuildingHandler) {
+                try {
+                    // looks like AsyncPagedListDiffer in 1.x ignores the config.
+                    // Reflection to the rescue.
+                    val mainThreadExecutorField =
+                        AsyncPagedListDiffer::class.java.getDeclaredField("mMainThreadExecutor")
+                    mainThreadExecutorField.isAccessible = true
+                    mainThreadExecutorField.set(this, Executor {
+                        modelBuildingHandler.post(it)
+                    })
+                } catch (t: Throwable) {
+                    val msg = "Failed to hijack update handler in AsyncPagedListDiffer." +
+                            "You can only build models on the main thread"
+                    Log.e("PagedListModelCache", msg, t)
+                    throw IllegalStateException(msg, t)
+                }
+            }
         }
-        rebuildCallback()
     }
-  }
-
-  private val asyncDiffer = @SuppressLint("RestrictedApi")
-  object : AsyncPagedListDiffer<T>(
-      updateCallback,
-      AsyncDifferConfig.Builder<T>(
-          itemDiffCallback
-      ).also {builder ->
-        if (diffExecutor != null) {
-          builder.setBackgroundThreadExecutor(diffExecutor)
+
+    @Synchronized
+    fun submitList(pagedList: PagedList<T>?) {
+        inSubmitList = true
+        asyncDiffer.submitList(pagedList)
+        inSubmitList = false
+    }
+
+    @Synchronized
+    fun getModels(): List<EpoxyModel<*>> {
+        val currentList = asyncDiffer.currentList ?: emptyList<T>()
+
+        // The first time models are built the EpoxyController does so synchronously, so that
+        // the UI can be ready immediately. To avoid concurrent modification issues with the PagedList
+        // and model cache we can't allow that first build to touch the cache.
+        if (Looper.myLooper() != modelBuildingHandler.looper) {
+            val initialModels = currentList.mapIndexed { position, item ->
+                modelBuilder(position, item)
+            }
+
+            // If the paged list still hasn't changed then we can populate the cache
+            // with the models we built to avoid needing to rebuild them later.
+            modelBuildingHandler.post {
+                setCacheValues(currentList, initialModels)
+            }
+
+            return initialModels
+        }
+
+        (0 until modelCache.size).forEach { position ->
+            if (modelCache[position] != null) {
```

Comment:
not sure maybe i wrot ethis code but i think it is easier to read if we just say

if (modelCache[position] == null) {
    // call build
}

to avoid return@foreach

or if we want to get kotlin fancy, we can do:

modelCache[position] ?: modelBuilder(position, list[position].also {  modelCache[position] = it}